### PR TITLE
Add Doio KB04

### PR DESCRIPTION
**Pull request by adophoxia**
## Description

As title. This was ironically only in the `vial_qmk` fork and not in here, so I've decided to "backport" it to upstream QMK.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* N/A

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/doio/kb04/keymaps/via/keymap.c [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        KC_TRNS,     KC_TRNS,    KC_TRNS,    KC_TRNS,     KC_TRNS
+    ),
+};
+
+#ifdef ENCODER_MAP_ENABLE
+const uint16_t PROGMEM encoder_map[][NUM_ENCODERS][2] = {
+    [_LAY0] = { ENCODER_CCW_CW(KC_VOLD, KC_VOLU) },
+    [_LAY1] = { ENCODER_CCW_CW(KC_TRNS, KC_TRNS) },
+    [_LAY2] = { ENCODER_CCW_CW(KC_TRNS, KC_TRNS) },
+    [_LAY3] = { ENCODER_CCW_CW(KC_TRNS, KC_TRNS) },
```
Comment (zvecr) [inline keyboards/doio/kb04/keymaps/via/keymap.c:56]:
```suggestion
const uint16_t PROGMEM encoder_map[][NUM_ENCODERS][NUM_DIRECTIONS] = {
    [_LAY0] = { ENCODER_CCW_CW(KC_VOLD, KC_VOLU) },
    [_LAY1] = { ENCODER_CCW_CW(KC_TRNS, KC_TRNS) }
```

File: keyboards/doio/kb04/keymaps/via/keymap.c [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    ),
+    [_LAY1] = LAYOUT(
+        KC_TRNS,     RGB_HUD,    RGB_TOG,    RGB_MOD,     KC_TRNS
+    ),
+    [_LAY2] = LAYOUT(
+        KC_TRNS,     KC_TRNS,    KC_TRNS,    KC_TRNS,     KC_TRNS
+    ),
+    [_LAY3] = LAYOUT(
+        KC_TRNS,     KC_TRNS,    KC_TRNS,    KC_TRNS,     KC_TRNS
+    ),
```
Comment (zvecr) [inline keyboards/doio/kb04/keymaps/via/keymap.c:48]:
```suggestion
    )
```

File: keyboards/doio/kb04/keymaps/via/keymap.c [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+ 
+ #include QMK_KEYBOARD_H
+
+enum layer_names {
+    _LAY0,
+    _LAY1,
+    _LAY2,
+    _LAY3
```
Comment (zvecr) [inline keyboards/doio/kb04/keymaps/via/keymap.c:23]:
```suggestion
    _LAY1
```

File: keyboards/doio/kb04/keymaps/default/keymap.c [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    [_LAY0] = LAYOUT(
+        MO(_LAY1),   KC_MPRV,    KC_MPLY,    KC_MNXT,     KC_MUTE
+    ),
+    [_LAY1] = LAYOUT(
+        KC_TRNS,     RGB_HUD,    RGB_TOG,    RGB_MOD,     KC_TRNS
+    )
+};
+
+#ifdef ENCODER_MAP_ENABLE
+const uint16_t PROGMEM encoder_map[][NUM_ENCODERS][2] = {
```
Comment (zvecr) [inline keyboards/doio/kb04/keymaps/default/keymap.c:44]:
```suggestion
const uint16_t PROGMEM encoder_map[][NUM_ENCODERS][NUM_DIRECTIONS] = {
```

File: keyboards/doio/kb04/info.json [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+            "solid_reactive_nexus": true,
+            "solid_reactive_multinexus": true,
+            "splash": true,
+            "multisplash": true,
+            "solid_splash": true,
+            "solid_multisplash": true
+        },
+        "default": {
+            "animation": "cycle_left_right"
+        },
```
Comment (zvecr) [inline keyboards/doio/kb04/info.json:77]:
```suggestion
```

File: keyboards/doio/kb04/info.json [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+            {"pin_a": "B5", "pin_b": "B6"}
+        ]
+    },
+    "features": {
+        "bootmagic": true,
+        "command": false,
+        "console": false,
+        "encoder": true,
+        "extrakey": true,
+        "lto": true,
```
Comment (zvecr) [inline keyboards/doio/kb04/info.json:18]:
```suggestion
```

File: keyboards/doio/kb04/keymaps/via/rules.mk [hunk 7]
```diff
@@ -0,0 +1,4 @@
+VIA_ENABLE = yes
+
+# Encoder enabled
```
Comment (zvecr) [inline keyboards/doio/kb04/keymaps/via/rules.mk:3]:
```suggestion
```

File: keyboards/doio/kb04/keymaps/default/rules.mk [hunk 8]
```diff
@@ -0,0 +1,2 @@
+# Encoder enabled
```
Comment (zvecr) [inline keyboards/doio/kb04/keymaps/default/rules.mk:1]:
```suggestion
```

File: keyboards/doio/kb04/info.json [hunk 9]
```diff
@@ -0,0 +1,113 @@
+{
+    "manufacturer": "DOIO",
+    "keyboard_name": "KB04-01",
+    "maintainer": "filmstarr",
+    "bootloader": "stm32duino",
+    "diode_direction": "COL2ROW",
```
Comment (zvecr) [inline keyboards/doio/kb04/info.json:6]:
```suggestion
    "build": {
        "lto": true
    },
    "diode_direction": "COL2ROW",
```

File: keyboards/doio/kb04/info.json [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+            {"matrix": [0, 2], "x": 149, "y": 0, "flags": 1},
+            {"matrix": [0, 3], "x": 224, "y": 0, "flags": 1},
+            {"matrix": [0, 4], "x": 112, "y": 64, "flags": 1},
+            {"matrix": [0, 5], "x": 112, "y": 64, "flags": 1},
+            {"matrix": [0, 6], "x": 112, "y": 64, "flags": 1},
+            {"matrix": [0, 7], "x": 112, "y": 64, "flags": 1}
+        ],
+        "max_brightness": 200,
+        "sleep": true,
+        "speed_steps": 30
```
Comment (zvecr) [inline keyboards/doio/kb04/info.json:91]:
```suggestion
        "sleep": true
```

**Review by zvecr — CHANGES_REQUESTED** [review]

File: keyboards/doio/kb04/readme.md [hunk 11]
```diff
@@ -0,0 +1,27 @@
+# doio/kb04
+
+![doio/kb04](https://cdn.shopify.com/s/files/1/0463/5980/6101/products/4_f40396cc-ea72-47a5-887f-bbeeb4023b11_1100x.png?v=1654787686)
```
Comment (waffle87) [inline keyboards/doio/kb04/readme.md:3]: This image should be rehosted on Imgur.

Comment (adophoxia) [reply keyboards/doio/kb04/readme.md]: Added in https://github.com/qmk/qmk_firmware/pull/22754/commits/545dca3f9e648d561353514d0b0f196b4dbfdf67.

File: keyboards/doio/kb04/info.json [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+        "command": false,
+        "console": false,
+        "encoder": true,
+        "extrakey": true,
+        "mousekey": true,
+        "nkro": false,
+        "rgb_matrix": true
+    },
+    "matrix_pins": {
+        "cols": ["B14", "B13", "B12", "B0", "A7", "A9", "A8", "A6"],
```
Comment (zvecr) [inline keyboards/doio/kb04/info.json:26]: 8 cols defined, mapped in rgb_matrix, but layout macro only exposes 5?

File: keyboards/doio/kb04/info.json [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+        "pin": "A10"
+    },
+    "layouts": {
+        "LAYOUT": {
+            "layout": [
+                {"label": "1!", "matrix": [0, 0], "x": 0, "y": 0},
+                {"label": "2@", "matrix": [0, 1], "x": 1, "y": 0},
+                {"label": "3#", "matrix": [0, 2], "x": 2, "y": 0},
+                {"label": "4$", "matrix": [0, 3], "x": 3, "y": 0},
+                {"label": "Encoder", "matrix": [0, 4], "x": 4.75, "y": 0}
```
Comment (zvecr) [inline keyboards/doio/kb04/info.json:107]: Does not match physical layout
```suggestion
                {"label": "Encoder", "matrix": [0, 4], "x": 1.5, "y": 1}
```

File: keyboards/doio/kb04/info.json (hunk 12 above)
Comment (adophoxia) [reply keyboards/doio/kb04/info.json]: Resolved in https://github.com/qmk/qmk_firmware/pull/22754/commits/69cd45e0d4c866bc98ba9d83d5d4caf798a129ab.

**Review by drashna — APPROVED** [review]

**Review by fauxpark — APPROVED** [review]

File: keyboards/doio/kb04/info.json [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+        "driver": "ws2812",
+        "layout": [
+            {"matrix": [0, 0], "x": 0, "y": 0, "flags": 1},
+            {"matrix": [0, 1], "x": 75, "y": 0, "flags": 1},
+            {"matrix": [0, 2], "x": 149, "y": 0, "flags": 1},
+            {"matrix": [0, 3], "x": 224, "y": 0, "flags": 1},
+            {"matrix": [0, 4], "x": 112, "y": 64, "flags": 1},
+            {"matrix": [0, 5], "x": 112, "y": 64, "flags": 1},
+            {"matrix": [0, 6], "x": 112, "y": 64, "flags": 1},
+            {"matrix": [0, 7], "x": 112, "y": 64, "flags": 1}
```
Comment (fauxpark) [inline keyboards/doio/kb04/info.json:86]: LED config is incorrect.

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/doio/kb04/info.json (hunk 14 above)
Comment (adophoxia) [reply keyboards/doio/kb04/info.json]: If this was referring to
```json
{"matrix": [0, 4], "x": 112, "y": 64, "flags": 1},
{"matrix": [0, 5], "x": 112, "y": 64, "flags": 1},
{"matrix": [0, 6], "x": 112, "y": 64, "flags": 1},
{"matrix": [0, 7], "x": 112, "y": 64, "flags": 1}
```
still being there from when the imaginary pins were removed, that was resolved in https://github.com/qmk/qmk_firmware/pull/22754/commits/a03e4a27cc3b9b332011f2cbaaf1b0e654b40a32 and https://github.com/qmk/qmk_firmware/pull/22754/commits/b0f79e2255f682f257718338665443e9243a00f9, unless of course, you meant something else.

Comment (fauxpark) [reply keyboards/doio/kb04/info.json]: image
🤷

Comment (adophoxia) [reply keyboards/doio/kb04/info.json]: Just compiled both `default` and `via`  =just now and they were successful. QMK CI must be playing catch up now.

**Review by fauxpark — APPROVED** [review]

File: keyboards/doio/kb04/info.json [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+        "rgb_matrix": true
+    },
+    "matrix_pins": {
+        "cols": ["B14", "B13", "B12", "B0", "A7"],
+        "rows": ["B3"]
+    },
+    "processor": "STM32F103",
+    "rgb_matrix": {
+        "animations": {
+            "solid_color": true,
```
Comment (drashna) [inline keyboards/doio/kb04/info.json:32]:
```suggestion
```

**Pull request merged**